### PR TITLE
Add support for STM32F412ZG

### PR DESCRIPTION
**Pull request by vaussard**
This adds support for STM32F412ZG, as well as the Nucleo STM32F412 board.

**Note 1**: I encountered some consistency weirdness in the stm32cube HAL. While most STM32 MCUs are grouped by SRAM/Flash size in the HAL (for instance STM32F401xC and STM32F401xE) or are size/package-agnostic (like the STM32F407xx), the STM32F412 is grouped by package size (STM32F412Cx, STM32F412Zx, STM32F412Vx and STM32F412Rx). The STM32F410 is the only other example of such design choice...

This does not play well with current implementation in Zephyr, as we define variants by SRAM/Flash size, not by package. This is important when defining CONFIG_SOC, which will configure the HAL at compile time.

So I chose to focus only on the ZG variant, because I need to know both the SRAM/Flash size for Zephyr and the package variant for the HAL :-(

**Note 2**: Following the recent DTS rework by Erwan (see commit 607fd149ece4ce64f66dc063d416a973092b082d), I tried to apply the same logic. But I must confess that I do not understand why, for example, stm32f407.dtsi includes stm32f411.dtsi, which in turn includes stm32f401.dtsi, which finally includes stm32f4.dtsi. Why such chaining? Is there really such a parent-child relation inside the IPs?

So my stm32f412.dtsi includes stm32f411.dtsi which results in the right number of UARTs, but I have no idea if this is right choice or not :-) Any advice?

### CODE REVIEW

**Comment by erwango** [comment]
Hi Florian,

Thanks for this. Quick feedback about your notes:
1) Indeed, this kind of partitioning is a bit odd. Looking to stm32f412 family, it appears there is one package variant with 3 UARTs (stm32f412Cx), while all other variants have 4 UARTs. This is a bit unusual, since package variants are usually "iso-feature". This is why cube specified different variants for same package. Not sure there is anything to fix here. Thanks for having taking care of it.

2)Actually, inside a STM32 series, a new SoC is most of the time a derivate, with few variations, of some other member of the same series (this is done to reduce design costs). Hence inheritance actually exists. Besides, 2 child SoCs can derivate from same parent SoC with some IPs  incompatibles from one another. Hence having a common dts file with all IPs is not possible. But we can have a minimum subset.
I should have caught this earlier, and explained it, when you initially pushed your commit.

Side note: we (ST) target to have common dts files for Linux, Zephyr and other projects using device tree for STM32, hence we should stick to some general rules, in order to keep compatibility across projects. This means it is still possible to have evolution in dts files due to alignment with other projects (don't be suprised).

My advice: I usually compare soc files (stm32f411xe.h vs stm32f412vg.h for instance) with an appropriate tool to see inheritance links between SoCs. Doing this, it appears stm32f413 could be a derivate of stm32f412 :-)

That being said, I'll try to review your PRs before end of week.

Erwan

**Comment by vaussard** [comment]
Hi Erwan,

Thank you for your explanations!

1. Yes I also saw the difference in number of declared IPs. I wondered if one could just use the include from the biggest package in all cases. It probably depends on the result of a memory access to an unimplemented IP.

2. Ok, it makes sense. I am just a bit worried that it will make reading the DTS difficult if one has to go through such a long include chain (stm32f413.dtsi -> stm32f412.dtsi -> stm32f411.dtsi -> stm32f401.dtsi -> stm32f4.dtsi). 

   I am not sure if it will be easy to merge such a scheme in Linux DTS. But I believe that such a question has not yet reached Linux, because most SoC have fewer variants and it is easier to come with a common .dtsi for the family. Anyway I am getting out of topic here...

Cheers,
Florian

File: boards/arm/nucleo_f412zg/nucleo_f412zg_defconfig [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+CONFIG_SOC_SERIES_STM32F4X=y
+CONFIG_SOC_STM32F412ZG=y
+# 96MHz system clock (highest value to get a precise USB clock)
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=96000000
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
+
+# Enable MPU
+CONFIG_STM32_ARM_MPU_ENABLE=y
+
+# enable USART2 - passthrough to STLINK v2 connector
```
Comment (erwango) [inline boards/arm/nucleo_f412zg/nucleo_f412zg_defconfig]: USART3

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_f412zg/nucleo_f412zg_defconfig (hunk 1 above)
Comment (vaussard) [inline boards/arm/nucleo_f412zg/nucleo_f412zg_defconfig]: Good catch, thank you!

Comment (vaussard) [inline boards/arm/nucleo_f412zg/nucleo_f412zg_defconfig]: Fixed!

File: boards/arm/nucleo_f412zg/doc/nucleo412zg.rst [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+     :align: center
+     :height: 540px
+     :alt: Nucleo F412ZG Morpho connectors (left)
+.. image:: img/nucleo_f412zg_morpho_right.png
+     :width: 720px
+     :align: center
+     :height: 540px
+     :alt: Nucleo F412ZG Morpho connectors (right)
+
+For mode details please refer to `STM32 Nucleo-144 board User Manual`_.
```
Comment (dbkinder) [inline boards/arm/nucleo_f412zg/doc/nucleo412zg.rst]: s/mode/more/

**Review by dbkinder — CHANGES_REQUESTED** [review]
one little typo in the doc, but otherwise LTGM

File: boards/arm/nucleo_f412zg/doc/nucleo412zg.rst (hunk 2 above)
Comment (vaussard) [inline boards/arm/nucleo_f412zg/doc/nucleo412zg.rst]: Thank you for spotting this typo. It is fixed.

**Review by dbkinder — APPROVED** [review]

**Comment by galak** [comment]
@erwango you good with the latest version?

**Review by erwango — APPROVED** [review]

**Comment by vaussard** [comment]
Fixed the merge conflict.

**Pull request merged**